fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolHandlers.java

```diff
@@ -0,0 +1,343 @@
+					}
+				}
+			}
+
+			return ActionResult.PASS;
+		}
+
+		@Override
+		public Float getMiningSpeedMultiplier(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
+			if (!(stack.getItem() instanceof DynamicAttributeTool)) {
+				float multiplier = stack.getItem().getMiningSpeed(stack, state);
+				if (multiplier != 1f) return multiplier;
+			}
+
+			return null;
+		}
+	}
+
+	/**
+	 * This handler handles items that are not a subclass of {@link DynamicAttributeTool} by
+	 * using the vanilla {@link Item#isEffectiveOn(BlockState)}.
+	 *
+	 * <p>Only applicable to blocks that are vanilla or share the material that is handled by their vanilla tool.</p>
+	 */
+	private static class VanillaToolsVanillaBlocksToolHandler implements ToolManager.ToolHandler {
+		@Override
+		public ActionResult isEffectiveOn(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
+			return !(stack.getItem() instanceof DynamicAttributeTool) && (stack.getItem().isEffectiveOn(state) || stack.getItem().getMiningSpeed(stack, state) != 1f) ? ActionResult.SUCCESS : ActionResult.PASS;
+		}
+
+		@Override
+		public Float getMiningSpeedMultiplier(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
+			if (!(stack.getItem() instanceof DynamicAttributeTool)) {
+				float miningSpeed = stack.getItem().getMiningSpeed(stack, state);
+
+				if (miningSpeed != 1f) {
+					return miningSpeed;
+				}
+			}
+
+			return null;
+		}
+	}
+
+	/**
+	 * This handler handles items that are a subclass of {@link DynamicAttributeTool} by using the
+	 * vanilla {@link Item#isEffectiveOn(BlockState)} with a custom fake tool material to use the mining level
+	 * from {@link DynamicAttributeTool#getMiningLevel(Tag, BlockState, ItemStack, LivingEntity)}.
+	 *
+	 * <p>Only applicable to blocks that are vanilla or share the material that is handled by their vanilla tool.</p>
+	 */
+	private static class ModdedMiningToolsVanillaBlocksToolHandler implements ToolManager.ToolHandler {
+		private final FakeAdaptableToolMaterial fakeMaterial = new FakeAdaptableToolMaterial();
+		private final ImmutableList<ToolItem> vanillaItems;
+
+		private ModdedMiningToolsVanillaBlocksToolHandler(ImmutableList<ToolItem> vanillaItems) {
+			this.vanillaItems = vanillaItems;
+		}
+
+		private ToolItem getVanillaItem(int miningLevel) {
+			if (miningLevel < 0) return vanillaItems.get(0);
+			if (miningLevel >= vanillaItems.size()) return vanillaItems.get(vanillaItems.size() - 1);
+			return vanillaItems.get(miningLevel);
+		}
+
+		@Override
+		public ActionResult isEffectiveOn(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
+			if (stack.getItem() instanceof DynamicAttributeTool) {
+				int miningLevel = ((DynamicAttributeTool) stack.getItem()).getMiningLevel(tag, state, stack, user);
+				if (miningLevel < 0) return ActionResult.PASS;
+				ToolItem vanillaItem = getVanillaItem(miningLevel);
+				ToolMaterial tempMaterial = vanillaItem.getMaterial();
```

Comment:
Messing with the material shouldn't be really necessary anymore, I'd rely on the retrieved vanilla tool as is. Vanilla shouldn't have any blocks that are incompatible with its own tools (=materials)

Reply:
What if we have tools that have mining levels outside 0 to 3? Or larger than level 1 axes?

Reply:
The next-best vanilla tool should suffice for vanilla blocks unless I am missing something. Even if the modded tool has level 20, a vanilla diamond tool should still be just as effective considering vanilla won't have level 20 blocks but at worst obsidian.

Reply:
True, I can do that for vanilla blocks